mivideo/mivideo_extract.py

```diff
@@ -67,62 +85,246 @@ def _readTableLastTime(self, tableName: str, tableColumnName: str) -> Union[date
+        kMedia = KalturaMediaService(kClient)
+
+        lastTime: Union[datetime, None] = (
+            self._readTableLastTime(tableName, 'created_at'))
+
+        if (lastTime):
+            localLogger.info(
+                f'Last time found in table: "{lastTime.isoformat()}"')
+        else:
+            lastTime = datetime.fromisoformat(self.defaultLastTimestamp)
+            localLogger.info('Last time not found in table; '
+                             f'using default time: "{lastTime.isoformat()}"')
+
+        createdAtTimestamp: float = lastTime.timestamp()
+
+        kFilter = KalturaMediaEntryFilter()
+        kFilter.createdAtGreaterThanOrEqual = createdAtTimestamp
+        kFilter.categoriesFullNameIn = 'Canvas_UMich'
+        kFilter.orderBy = KalturaMediaEntryOrderBy.CREATED_AT_ASC
+
+        kPager = KalturaFilterPager()
+        kPager.pageSize = 500  # 500 is maximum
+        kPager.pageIndex = 1
+
+        results: Sequence[KalturaMediaEntry] = None
+        lastCreatedAtTimestamp: Union[float, int] = createdAtTimestamp
+        lastId: Union[str, None] = None
+        numberResults: int = 0
+        queryPageNumber: int = kPager.pageIndex  # for logging purposes
+        totalNumberResults: int = numberResults  # for logging purposes
+
+        while True:
+            try:
+                results = kMedia.list(kFilter, kPager).objects
+            except KalturaException as kException:
+                if (KALTURA_MAX_MATCHES_ERROR in kException.args):
+                    # set new filter timestamp, reset pager to page 1, then continue
+                    kFilter.createdAtGreaterThanOrEqual = lastCreatedAtTimestamp
+                    localLogger.debug(
+                        f'New filter timestamp: ({kFilter.createdAtGreaterThanOrEqual})')
+
+                    # to avoid dupes, also filter out the last ID returned by previous query
+                    # because Kaltura compares createdAt greater than *or equal* to timestamp
+                    kFilter.idNotIn = lastId
+                    kPager.pageIndex = 1
+                    continue
+
+                localLogger.info(f'Other Kaltura API error: "{kException}"')
+                break
+
+            numberResults = len(results)
+            localLogger.debug(
+                f'Query page ({queryPageNumber}); number of results: ({numberResults})')
+
+            if (numberResults > 0):
+                resultDictionaries: Sequence[Dict] = tuple(r.__dict__ for r in results)
+
+                creationData: pd.DataFrame = self._makeCreationData(resultDictionaries)
+
+                creationData.to_sql(
+                    tableName, self.appDb.engine, if_exists='append', index=False)
+
+                courseData: pd.DataFrame = self._makeCourseData(resultDictionaries)
+
+                courseData.to_sql('mivideo_media_courses', self.appDb.engine, if_exists='append',
+                                  index=False, method=self._queryRunner)
```

Comment:
Yeah, simply that I thought writing progressively would be better than waiting to get everything first, risking data loss and large transactions.  Since each page of results is independent, there's no reason to keep them together.  I also thought that it would make the calls to Pandas **_slightly_** simpler; always creating a new dataframe rather than creating one the first time, then appending thereafter.

Reply:
Okay, yeah, if the data is always different on each page (i.e. you don't need to drop duplicates ever across pages), this approach should work fine. Just curious as to your thinking. I kinda go back and forth about whether it would be worse to have an incomplete dataset (it fails halfway through) or have no data at all (it fails anytime before a final insertion).

Reply:
I see.  In this case, it's fine to have an incomplete dataset.  The query always looks for anything newer than the last timestamp in the DB.  If something causes this program to fail part way through processing the API results, then the next time the program runs successfully, it will continue with the last timestamp found in the DB.